File: src/0166-Fraction-to-Recurring-Decimal/0166.cpp
class Solution 
{
public:
    string fractionToDecimal(int numerator, int denominator) 
    {
        long long n = numerator, d = denominator;
        bool flag = true;
        if (n < 0) flag = !flag, n = -n;
        if (d < 0) flag = !flag, d = -d;
        string res = to_string(n/d);
        n %= d;
        if (n == 0) 
        {
            if (!flag and res != "0") return "-" + res;
            return res;
        }
        
        res += '.';
        unordered_map<long long, int> m;
        while (n) 
        {
            if (m.count(n)) 
            {
                res = res.substr(0, m[n]) + "(" + res.substr(m[n]) +")";
                break;
            }
            else m[n] = res.size();
            n *= 10ll;
            res += to_string(n/d);
            n %= d;
        }
        return flag ? res : "-" + res;
    }
};

File: src/0166-Fraction-to-Recurring-Decimal/0166.py
class Solution:
    def fractionToDecimal(self, n, d):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        flag = True
        if n < 0:
            flag = not flag
            n = -n
        
        if d < 0:
            flag = not flag
            d = -d
            
        res = str(n // d)
        n %= d
        if n == 0:
            if not flag and res != "0":
                return "-" + res
            return res
        
        res += '.'
        m = dict()
        while n:
            if n in m:
                res = res[0:m[n]] + "(" + res[m[n]:] + ")"
                break
            else:
                m[n] = len(res)
                
            n *= 10
            res += str(n // d)
            n %= d
            
        return res if flag else "-" + res